File: go/solutions/0020_valid_parentheses/valid_parentheses.go
package valid_parentheses

// 解法一：使用栈+哈希表
// Time: O(n), Space: O(n)
func isValid(s string) bool {
	pairs := map[byte]byte{
		')': '(',
		']': '[',
		'}': '{',
	}

	stack := []byte{}
	for i := 0; i < len(s); i++ { // 注意 range 返回的是 rune，这个返回的是 byte
		if s[i] == '(' || s[i] == '[' || s[i] == '{' {
			stack = append(stack, s[i])
		} else {
			n := len(stack)
			if n == 0 || stack[n-1] != pairs[s[i]] {
				return false
			}
			stack = stack[:n-1]
		}
	}

	return len(stack) == 0
}

File: python/solutions/0020_valid_parentheses/valid_parentheses.py
class Solution:
    def isValidStack(self, s: str) -> bool:
        if len(s) % 2 == 1:
            return False

        pairs = {
            ")": "(",
            "]": "[",
            "}": "{",
        }
        stack: list[str] = []
        for ch in s:
            if ch in pairs:
                if not stack or stack.pop() != pairs[ch]:
                    return False
            else:
                stack.append(ch)

        return not stack
